domain/consensus/utils/consensushashing/calculate_signature_hash.go

```diff
@@ -0,0 +1,196 @@
+	prevScriptPublicKey *externalapi.ScriptPublicKey, hashType SigHashType, reusedValues *SighashReusedValues) (
+	*externalapi.DomainHash, error) {
+
+	hashWriter := hashes.NewTransactionSigningHashWriter()
+	infallibleWriteElement(hashWriter, tx.Version)
+
+	previousOutputsHash := getPreviousOutputsHash(tx, hashType, reusedValues)
+	infallibleWriteElement(hashWriter, previousOutputsHash)
+
+	sequencesHash := getSequencesHash(tx, hashType, reusedValues)
+	infallibleWriteElement(hashWriter, sequencesHash)
+
+	hashOutpoint(hashWriter, txIn)
+
+	infallibleWriteElement(hashWriter, prevScriptPublicKey.Version)
+	infallibleWriteElement(hashWriter, prevScriptPublicKey.Script)
+
+	infallibleWriteElement(hashWriter, txIn.UTXOEntry.Amount())
+
+	infallibleWriteElement(hashWriter, txIn.Sequence)
+
+	outputsHash := getOutputsHash(tx, inputIndex, hashType, reusedValues)
+	infallibleWriteElement(hashWriter, outputsHash)
+
+	infallibleWriteElement(hashWriter, tx.LockTime)
+
+	infallibleWriteElement(hashWriter, tx.SubnetworkID)
+	infallibleWriteElement(hashWriter, tx.Gas)
+
+	payloadHash := getPayloadHash(tx, reusedValues)
+	infallibleWriteElement(hashWriter, payloadHash)
+
+	infallibleWriteElement(hashWriter, uint32(hashType))
+
+	return hashWriter.Finalize(), nil
+}
+
+func getPreviousOutputsHash(tx *externalapi.DomainTransaction, hashType SigHashType, reusedValues *SighashReusedValues) *externalapi.DomainHash {
+	if hashType.isSigHashAnyOneCanPay() {
+		return externalapi.NewZeroHash()
+	}
+
+	if reusedValues.previousOutputsHash == nil {
+		hashWriter := hashes.NewTransactionSigningHashWriter()
+		for _, txIn := range tx.Inputs {
+			hashOutpoint(hashWriter, txIn)
+		}
+		reusedValues.previousOutputsHash = hashWriter.Finalize()
+	}
+
+	return reusedValues.previousOutputsHash
+}
+
+func getSequencesHash(tx *externalapi.DomainTransaction, hashType SigHashType, reusedValues *SighashReusedValues) *externalapi.DomainHash {
+	if hashType.isSigHashSingle() || hashType.isSigHashAnyOneCanPay() || hashType.isSigHashNone() {
+		return externalapi.NewZeroHash()
+	}
+
+	if reusedValues.sequencesHash == nil {
+		hashWriter := hashes.NewTransactionSigningHashWriter()
+		for _, txIn := range tx.Inputs {
+			infallibleWriteElement(hashWriter, txIn.Sequence)
+		}
+		reusedValues.sequencesHash = hashWriter.Finalize()
+	}
+
+	return reusedValues.sequencesHash
+}
+
+func getOutputsHash(tx *externalapi.DomainTransaction, inputIndex int, hashType SigHashType, reusedValues *SighashReusedValues) *externalapi.DomainHash {
+	// SigHashNone: return zero-hash
+	if hashType.isSigHashNone() {
+		return externalapi.NewZeroHash()
+	}
+
+	// SigHashSingle: If the relevant output exists - return it's hash, otherwise return zero-hash
+	if hashType.isSigHashSingle() {
+		if inputIndex >= len(tx.Outputs) {
+			return externalapi.NewZeroHash()
+		}
+		hashWriter := hashes.NewTransactionSigningHashWriter()
+		hashTxOut(hashWriter, tx.Outputs[inputIndex])
+		return hashWriter.Finalize()
```

Comment:
I think you should update reusedValues.outputsHash here and only return later.

Reply:
this outputsHash isn't cached, because if it's single then each input will have its own outputsHash, you can't cache them between inputs